node/src/reactor/participating.rs

```diff
@@ -539,20 +540,150 @@ impl reactor::Reactor for Reactor {
+
+        // If we have a latest block header, we joined via the joiner reactor.
+        //
+        // If not, run genesis or upgrade and construct a switch block, and use that for the latest
+        // block header.
+        let latest_block_header = if let Some(latest_block_header) = maybe_latest_block_header {
+            initial_era = latest_block_header.next_block_era_id();
+            latest_block_header
+        } else {
+            let initial_pre_state;
+            let finalized_block;
+
+            let chainspec = chainspec_loader.chainspec();
+            match chainspec_loader
+                .chainspec()
+                .protocol_config
+                .activation_point
+            {
+                ActivationPoint::Genesis(genesis_timestamp) => {
+                    // Check that no blocks exist in storage so we don't try to run genesis again on
+                    // an existing blockchain
+                    if let Some(first_block_header) = storage.read_block_header_by_height(1)? {
+                        return Err(Error::CannotRunGenesisOnPreExistingBlockchain {
+                            first_block_header: Box::new(first_block_header),
+                        });
+                    }
+                    let GenesisSuccess {
+                        post_state_hash,
+                        execution_effect,
+                    } = contract_runtime.commit_genesis(chainspec)?;
+                    info!("genesis chainspec name {}", chainspec.network_config.name);
+                    info!("genesis state root hash {}", post_state_hash);
+                    trace!(%post_state_hash, ?execution_effect);
+                    initial_pre_state = ExecutionPreState::new(
+                        post_state_hash.into(),
+                        0,
+                        Default::default(),
+                        Default::default(),
+                    );
+                    finalized_block = FinalizedBlock::new(
+                        BlockPayload::default(),
+                        Some(EraReport::default()),
+                        genesis_timestamp,
+                        EraId::from(0u64),
+                        0,
+                        PublicKey::System,
+                    );
+                    initial_era = EraId::from(1);
+                }
+                ActivationPoint::EraId(upgrade_era_id) => {
+                    let upgrade_block_header = storage
+                        .read_switch_block_header_by_era_id(upgrade_era_id.saturating_sub(1))?
+                        .ok_or(Error::NoSuchSwitchBlockHeaderForUpgradeEra { upgrade_era_id })?;
+                    // Unless it's an emergency upgrade, if there's a block higher than ours we will
+                    // over-write an existing blockchain if we don't exit immediately.
+                    if chainspec.protocol_config.last_emergency_restart
+                        == Some(upgrade_block_header.era_id())
```

Comment:
Is this correct? This seems to ask if we are currently in an emergency upgrade, not `unless it's an emergency upgrade`. Perhaps the comment should be "If we are **not** doing an emergency upgrade, and if there's a block higher than ours..."